In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn import preprocessing
from sklearn import metrics
import seaborn as sns
from sklearn import svm

In [2]:
train = pd.read_csv('train.csv')
test = pd.read_csv('test.csv')

In [3]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
PassengerId    891 non-null int64
Survived       891 non-null int64
Pclass         891 non-null int64
Name           891 non-null object
Sex            891 non-null object
Age            714 non-null float64
SibSp          891 non-null int64
Parch          891 non-null int64
Ticket         891 non-null object
Fare           891 non-null float64
Cabin          204 non-null object
Embarked       889 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [4]:
train.isna().any()

PassengerId    False
Survived       False
Pclass         False
Name           False
Sex            False
Age             True
SibSp          False
Parch          False
Ticket         False
Fare           False
Cabin           True
Embarked        True
dtype: bool

In [5]:
train = train.drop(['Name','Cabin','Ticket','PassengerId'], axis=1)

In [6]:
train = pd.get_dummies(train)
train.columns

Index(['Survived', 'Pclass', 'Age', 'SibSp', 'Parch', 'Fare', 'Sex_female',
       'Sex_male', 'Embarked_C', 'Embarked_Q', 'Embarked_S'],
      dtype='object')

In [7]:
#impute age column with median
train['Age'].fillna(train['Age'].median(),inplace = True)

In [8]:
train.isna().any()

Survived      False
Pclass        False
Age           False
SibSp         False
Parch         False
Fare          False
Sex_female    False
Sex_male      False
Embarked_C    False
Embarked_Q    False
Embarked_S    False
dtype: bool

In [9]:
x = train.drop(['Survived'],axis=1)
y = train['Survived']

In [10]:
x = preprocessing.scale(x)

from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=0.2,random_state=0)

In [11]:
clf = svm.SVC()
clf.fit(x_train,y_train)

SVC(C=1.0, break_ties=False, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma='scale', kernel='rbf',
    max_iter=-1, probability=False, random_state=None, shrinking=True,
    tol=0.001, verbose=False)

In [12]:
y_pred = clf.predict(x_test)

In [13]:
print("Accuracy:",round(metrics.accuracy_score(y_test,y_pred),3))
print("Precision:",round(metrics.precision_score(y_test,y_pred),3))
print("Recall:",round(metrics.recall_score(y_test,y_pred),3))
print("F1 score:",round(metrics.f1_score(y_test,y_pred),3))

Accuracy: 0.804
Precision: 0.793
Recall: 0.667
F1 score: 0.724


In [22]:
clf2 = svm.SVC(C=0.001,gamma=0.0001)
clf2.fit(x_train,y_train)
y_pred2 = clf2.predict(x_test)

print("Accuracy:",round(metrics.accuracy_score(y_test,y_pred2),3))
print("Precision:",round(metrics.precision_score(y_test,y_pred2),3))
print("Recall:",round(metrics.recall_score(y_test,y_pred2),3))
print("F1 score:",round(metrics.f1_score(y_test,y_pred2),3))

Accuracy: 0.637
Precision: 0.0
Recall: 0.0
F1 score: 0.0


C:\Miniconda\lib\site-packages\sklearn\metrics\_classification.py:1272: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [23]:
clf3 = svm.SVC(C=0.01,gamma=0.001)
clf3.fit(x_train,y_train)
y_pred3 = clf3.predict(x_test)

print("Accuracy:",round(metrics.accuracy_score(y_test,y_pred3),3))
print("Precision:",round(metrics.precision_score(y_test,y_pred3),3))
print("Recall:",round(metrics.recall_score(y_test,y_pred3),3))
print("F1 score:",round(metrics.f1_score(y_test,y_pred3),3))

Accuracy: 0.637
Precision: 0.0
Recall: 0.0
F1 score: 0.0


C:\Miniconda\lib\site-packages\sklearn\metrics\_classification.py:1272: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [24]:
clf4 = svm.SVC(C=0.5,gamma=0.01)
clf4.fit(x_train,y_train)
y_pred4 = clf4.predict(x_test)

print("Accuracy:",round(metrics.accuracy_score(y_test,y_pred4),3))
print("Precision:",round(metrics.precision_score(y_test,y_pred4),3))
print("Recall:",round(metrics.recall_score(y_test,y_pred4),3))
print("F1 score:",round(metrics.f1_score(y_test,y_pred4),3))

Accuracy: 0.732
Precision: 0.649
Recall: 0.569
F1 score: 0.607


In [25]:
clf5 = svm.SVC(C=1,gamma=0.1)
clf5.fit(x_train,y_train)
y_pred5 = clf5.predict(x_test)

print("Accuracy:",round(metrics.accuracy_score(y_test,y_pred5),3))
print("Precision:",round(metrics.precision_score(y_test,y_pred5),3))
print("Recall:",round(metrics.recall_score(y_test,y_pred5),3))
print("F1 score:",round(metrics.f1_score(y_test,y_pred5),3))

Accuracy: 0.821
Precision: 0.884
Recall: 0.585
F1 score: 0.704


In [26]:
clf6 = svm.SVC(C=2,gamma=0.25)
clf6.fit(x_train,y_train)
y_pred6 = clf6.predict(x_test)

print("Accuracy:",round(metrics.accuracy_score(y_test,y_pred6),3))
print("Precision:",round(metrics.precision_score(y_test,y_pred6),3))
print("Recall:",round(metrics.recall_score(y_test,y_pred6),3))
print("F1 score:",round(metrics.f1_score(y_test,y_pred6),3))

Accuracy: 0.804
Precision: 0.857
Recall: 0.554
F1 score: 0.673


In [28]:
clf7 = svm.SVC(C=5,gamma=0.5)
clf7.fit(x_train,y_train)
y_pred7 = clf7.predict(x_test)

print("Accuracy:",round(metrics.accuracy_score(y_test,y_pred7),3))
print("Precision:",round(metrics.precision_score(y_test,y_pred7),3))
print("Recall:",round(metrics.recall_score(y_test,y_pred7),3))
print("F1 score:",round(metrics.f1_score(y_test,y_pred7),3))

Accuracy: 0.777
Precision: 0.766
Recall: 0.554
F1 score: 0.643


In [29]:
clf8 = svm.SVC(C=5,gamma=1)
clf8.fit(x_train,y_train)
y_pred8 = clf8.predict(x_test)

print("Accuracy:",round(metrics.accuracy_score(y_test,y_pred8),3))
print("Precision:",round(metrics.precision_score(y_test,y_pred8),3))
print("Recall:",round(metrics.recall_score(y_test,y_pred8),3))
print("F1 score:",round(metrics.f1_score(y_test,y_pred8),3))

Accuracy: 0.788
Precision: 0.755
Recall: 0.615
F1 score: 0.678


# Tuning the model parameters by modifying the pipeline

In [15]:
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.svm import SVC
from sklearn.pipeline import Pipeline 

#Setting up a ml pipeline
pipeline = Pipeline([('clf', SVC(kernel = 'rbf', C = 1, gamma = 0.1))]) 
#Setting the parameters that shall be used during grid search
params = {'clf__C':(0.001, 0.01, 0.1, 0.5, 1, 2, 5), 
          'clf__gamma':(0.0001, 0.001, 0.01, 0.1, 0.25, 0.5, 0.75, 1)} 

#Declaring the grid search to create several SVM ML in the pipeline each with all the declared parameters
svm_grid_rbf = GridSearchCV(pipeline, params, n_jobs = -1,
                            cv = 3, verbose = 0, scoring='accuracy') 
#Fitting the grid search ML with the training data
svm_grid_rbf.fit(x_train, y_train) 

GridSearchCV(cv=3, error_score=nan,
             estimator=Pipeline(memory=None,
                                steps=[('clf',
                                        SVC(C=1, break_ties=False,
                                            cache_size=200, class_weight=None,
                                            coef0=0.0,
                                            decision_function_shape='ovr',
                                            degree=3, gamma=0.1, kernel='rbf',
                                            max_iter=-1, probability=False,
                                            random_state=None, shrinking=True,
                                            tol=0.001, verbose=False))],
                                verbose=False),
             iid='deprecated', n_jobs=-1,
             param_grid={'clf__C': (0.001, 0.01, 0.1, 0.5, 1, 2, 5),
                         'clf__gamma': (0.0001, 0.001, 0.01, 0.1, 0.25, 0.5,
                                        0.75, 1)},
  

In [16]:
#using .get_params() to identify the set of parameters that enabled the model to perform the best
best = svm_grid_rbf.best_estimator_.get_params() 

for k in sorted(params.keys()): 
    print(f'\t{k}: \t {round(best[k], 2)}')

	clf__C: 	 2
	clf__gamma: 	 0.1


In [17]:
#Setting up the ML with parameters based on the Grid Search
svc = SVC(C = 2, gamma = 0.10)
svc.fit(x_train, y_train)

SVC(C=2, break_ties=False, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma=0.1, kernel='rbf',
    max_iter=-1, probability=False, random_state=None, shrinking=True,
    tol=0.001, verbose=False)

In [18]:
y_predd = svc.predict(x_test)

In [20]:
#Assessing the median Accuracy Scores
print('Accuracy:', round(metrics.accuracy_score(y_test, y_predd), 3))
print( 'Precision:', round(metrics.precision_score(y_test, y_predd), 3))
print( 'Recall:', round(metrics.recall_score(y_test, y_predd), 3))
print( 'F1 Score:', round(metrics.f1_score(y_test, y_predd), 3))

Accuracy: 0.81
Precision: 0.818
Recall: 0.652
F1 Score: 0.726


# Conclusion and comparison

	SVC(C=1,gamma=0.1)	Logistic Regression(C=1)
Accuracy	82.10%	           73%
Precision	88.40%	           81.20%
Recall	    58.50%	           63%
F1 Score	70.40%	           71%
		
** The logistic Regression model appears to be more sensitive than the SVC. Given that the recall and precision is higher, it would mean this model is able to identify relevant instances better and show that they are actually better.		